### PR TITLE
First Implementation

### PR DESCRIPTION
**Pull request by WillB97**
- Provides a read-only filesystem containing the `astoria.json` file.
- Disables the serial terminal.
- Sets the two outer LEDs to the corner colour based the contents of the `astoria.json` file.
- Includes a heartbeat LED to show the MCU is still running.
- Uses a custom JSON library because CircuitPython doesn't include one for non-express devices and I felt like writing one.

### CODE REVIEW

File: src/json.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
+"""json.py
```
Comment (RealOrangeOne) [inline src/json.py:1]: Question: Was this custom written, or sourced from somewhere else?

File: src/boot.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+boot.py
+
+This file is executed before the USB connection is setup so we can alter what
+USB devices are created.
+"""
+import usb_cdc
+import storage
+
+# Make USB mount read-only (to host), this can only be done in boot.py
+storage.remount('/', readonly=False)
```
Comment (RealOrangeOne) [inline src/boot.py]: > Make USB mount read-only

`readonly=False`

Eh?

File: src/boot.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+# Set the name the mass storage appears as,
+# this can only be done while the storage is writable to the microcontroller
+m = storage.getmount("/")
+if m.label == 'CIRCUITPY':
```
Comment (RealOrangeOne) [inline src/boot.py]: Question: Any reason not to reset the label in all cases?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,51 @@
...
+An implementation of the [Metadata USBs](https://srobo.github.io/astoria/usage.html#metadata-usbs) in CircuitPython using an Adafruit Neo Trinkey.
+
+# Configuring a Neo Trinkey
+
+The Neo Trinkey does not come with the CircuitPython bootloader installed.
+
+To install the bootloader, plug the Trinkey and double press the reset button.
+All 4 LEDs should now be green and a TRINKEYBOOT drive should be visible, if they are red there is a problem with the USB connection.
+Download the CircuitPython bootloader from the [circuitpython website](https://circuitpython.org/board/neopixel_trinkey_m0/) and drag the uf2 file onto the TRINKEYBOOT drive.
+The Neo Trinkey will then reboot and a CiRCUITPY drive will appear.
```
Comment (RealOrangeOne) [inline README.md]: Question: "CiRCUITPY" or "CIRCUITPY"? (capital "I")

File: src/code.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+# Orange is modified here to make it more distinct from yellow
+corners = [0x00ff00, 0xff3300, 0xff00ff, 0xffff00]
+
+PIXELS = neopixel.NeoPixel(board.NEOPIXEL, 4)
+PIXELS.brightness = 0.02  # The LEDs are very bright
+
+
+def set_corner(corner, arena=None):
+    "Set LEDs 1/2 to the corner's colour, the LED set depends on the arena"
```
Comment (RealOrangeOne) [inline src/code.py:15]: Question: How are we intending to tell which device is which without plugging them into power? Looking at the photos, it doesn't look like there's much of a keyring attachment.

Comment (TomWheal) [reply src/code.py:15]: Heat shrink with a piece of card on the bottom side would work nicely and help protect the device

File: src/json.py (hunk 1 above)
Comment (WillB97) [reply src/json.py:1]: Custom written using the syntax diagrams from https://www.json.org/.

File: src/boot.py (hunk 3 above)
Comment (WillB97) [reply src/boot.py]: Setting the label is written to non-volatile memory so it consumes flash write cycles which should be avoided. Although the test case should be inverted.

File: README.md (hunk 4 above)
Comment (WillB97) [reply README.md]: My pinky got tired. Yes, it's all caps.

File: src/boot.py (hunk 2 above)
Comment (WillB97) [reply src/boot.py]: The virtual drive can only be writeable from one location (MCU or host), the read-only argument is from the perspective of the MCU.

Yes, the comment needs improvement.

File: demo/code.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+# Remember to update the colours in src/code.py if you change them here
+corners = [0x00ff00, 0xff3300, 0xff00ff, 0xffff00]
+
+PIXELS = neopixel.NeoPixel(board.NEOPIXEL, 4)
+PIXELS.brightness = 0.02  # The LEDs are very bright
+
+# Preview corner colours
+PIXELS[:] = corners
+while True:
+    time.sleep(1)
```
Comment (trickeydan) [inline demo/code.py:15]: Question: Is this file run at any point? `src/code.py` is the entry point and I can't see this file referenced elsewhere?

File: src/code.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+def get_zone_from_file(filename):
+    "Read the zone and arena from a metadata file, all LEDs are held blue if an error occurs"
+    try:
+        with open(filename) as fp:
+            config = fp.read()
+            data = json.loads(config)
+
+        return data['zone'], data['arena']
+    except (OSError, ValueError, IndexError) as e:
+        print(e)
```
Comment (trickeydan) [inline src/code.py:45]: Question: We have disabled the serial console, so does this print actually go anywhere?

File: src/code.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        # All LEDs are set blue if the mode file is invalid
+        PIXELS.fill(0x0000FF)
+        while True:
+            time.sleep(1)
+
+
+# LEDs set to corner colour
+set_corner(*get_zone_from_file('astoria.json'))
```
Comment (trickeydan) [inline src/code.py]:
```suggestion
zone, arena = get_zone_from_file('astoria.json')
set_corner(zone, arena)
```

**Review by trickeydan — COMMENTED** [review]
🎉 Good stuff!

There are probably a couple of things we could do to improve stuff here, but ultimately this is more than good enough for an initial implementation.

Not approved explicitly in lieu of my query around `demo/code.py`, but otherwise I'm happy with this!

File: demo/code.py (hunk 6 above)
Comment (WillB97) [reply demo/code.py:15]: Forgot part of the readme. The demo folder is an alternative (and separate) firmware to allow for testing the corner colours.

File: src/code.py (hunk 7 above)
Comment (WillB97) [reply src/code.py:45]: By default this goes nowhere but if you boot into safe mode you can run `import code` to see the error.

I'll add this to the readme

File: src/json.py (hunk 1 above)
Comment (WillB97) [reply src/json.py:1]: a8bd3104fdda2c74b4f76ed88574bc1bfffe9eb0

File: src/boot.py (hunk 2 above)
Comment (WillB97) [reply src/boot.py]: bef71f6b627e579cfcf261b9e205bf0b7bf3cf97

File: README.md (hunk 4 above)
Comment (WillB97) [reply README.md]: bef71f6b627e579cfcf261b9e205bf0b7bf3cf97

File: demo/code.py (hunk 6 above)
Comment (WillB97) [reply demo/code.py:15]: afbcf5d57f3a86c98ac6ab8ead619c77db20aa1f

File: src/code.py (hunk 7 above)
Comment (WillB97) [reply src/code.py:45]: afbcf5d57f3a86c98ac6ab8ead619c77db20aa1f

**Review by trickeydan — APPROVED** [review]

**Pull request merged**